src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/StoppingConditionByNumberOfGenerations.java

```diff
@@ -0,0 +1,26 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.randomized;
+
+public class StoppingConditionByNumberOfGenerations implements StoppingConditionForIterativeImprovement{
+	
+    protected final int generationThreshold;
+    protected int generations;
```

Comment:
This should not be a member variable in this class because any object of this class should be stateless so that we can use just one such object, exactly like we are using only one `IterativeImprovementBasedQueryOptimizer` object. In other words, calling the `readyToStop` method should not change the state of this object because we may (re)use this object for multiple queries to be optimized (perhaps even in parallel).

Therefore, after removing this member variable here, you need to extend the signature of the `readyToStop` method (not only here but also in the interface) by adding a parameter via which the current generation counter can be passed:
```java
public interface StoppingConditionForIterativeImprovement {
	boolean readyToStop( int numOfCurrentGeneration );
}
```